### PR TITLE
[website,crm,livechat,visitor] display visitors info on crm lead and handle lead merge

### PR DESCRIPTION
**Pull request by dbeguin**

### CODE REVIEW

File: addons/website_crm/models/crm_lead.py [hunk 1]
```diff
@@ -7,7 +7,39 @@
...
+                    WHERE l.id in %s
+                    GROUP BY l.id"""
+        self.env.cr.execute(sql, (tuple(self.ids),))
+        page_data = self.env.cr.dictfetchall()
+        mapped_data = {data['lead_id']: data['page_view_count'] for data in page_data}
+        for lead in self:
+            lead.visitor_page_count = mapped_data.get(lead.id, 0)
+
+    def action_redirect_to_page_views(self):
+        self.ensure_one()
```
Comment (tde-banana-odoo) [inline addons/website_crm/models/crm_lead.py]: Ensure one is not necessary, you would have the same action for a true multi record set if you want. isn't it ?

File: addons/website_crm/models/crm_lead.py [hunk 2]
```diff
@@ -7,7 +7,39 @@
...
+        visitors = self.visitor_ids
+        action = self.env.ref('website.website_visitor_page_action').read()[0]
+        action['domain'] = [('visitor_id', 'in', visitors.ids)]
+        # avoid grouping if only few records
```
Comment (tde-banana-odoo) [inline addons/website_crm/models/crm_lead.py]: Well we generally don't do that ... is it a hard requirement of the marketing team that cannot work without that top notch feature ?

File: addons/website_crm/models/website_visitor.py [hunk 3]
```diff
@@ -7,12 +7,19 @@
 class WebsiteVisitor(models.Model):
     _inherit = 'website.visitor'
 
-    lead_ids = fields.One2many('crm.lead', 'visitor_id', string='Leads')
+    lead_ids = fields.Many2many('crm.lead', string='Leads')
```
Comment (tde-banana-odoo) [inline addons/website_crm/models/website_visitor.py]: I have some questions here ...Visitor is a marketing oriented record. It is not obvious people have rights on leads. Previously computation of lead count would either crash, either give results depending on ACLS. Currently you always have results independently of your access rights.

I would
 * display lead_ids and lead_count only if people are at least sales something group as in all other modules;
 * remove the hand made sql and replace it by a real search depending on acls;
 * add some rights on visitor model in website_crm for lead people as it does not seem to be the case;

File: addons/website_crm_livechat/__manifest__.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    'category': 'Website',
+    'sequence': 54,
+    'summary': 'View livechat sessions for leads',
+    'version': '1.0',
+    'description': """ Adds a stat button on lead form view to access their livechat sessions.""",
+    'depends': ['website_crm', 'website_livechat'],
+    'data': [
+        'views/website_crm_lead_views.xml',
+    ],
+    'qweb': ['static/src/xml/*.xml'],
```
Comment (tde-banana-odoo) [inline addons/website_crm_livechat/__manifest__.py]: qweb is not necessary

File: addons/website_crm_livechat/models/crm_lead.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+from odoo import fields, models
+
+
+class Lead(models.Model):
+    _inherit = 'crm.lead'
+
+    visitor_sessions_count = fields.Integer('# Sessions', compute="_compute_visitor_sessions_count")
+
+    def _compute_visitor_sessions_count(self):
```
Comment (tde-banana-odoo) [inline addons/website_crm_livechat/models/crm_lead.py]: Missing triggers

File: addons/website_crm_livechat/models/crm_lead.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                    JOIN crm_lead l on l.id = lv.crm_lead_id
+                    WHERE l.id in %s
+                    GROUP BY l.id"""
+        self.env.cr.execute(sql, (tuple(self.ids),))
+        page_data = self.env.cr.dictfetchall()
+        mapped_data = {data['lead_id']: data['session_count'] for data in page_data}
+        for lead in self:
+            lead.visitor_sessions_count = mapped_data.get(lead.id, 0)
+
+    def action_redirect_to_livechat_sessions(self):
```
Comment (tde-banana-odoo) [inline addons/website_crm_livechat/models/crm_lead.py]: Same as above, not really necessary to ensure one it seems.

File: addons/website_crm_livechat/__manifest__.py (hunk 4 above)
Comment (dbeguin) [reply addons/website_crm_livechat/__manifest__.py]: indeed

File: addons/website_crm/models/crm_lead.py (hunk 2 above)
Comment (dbeguin) [reply addons/website_crm/models/crm_lead.py]: it is a @odoongr and @LNAodoo 'make it easy to read and more convenient to use in fonction of the case' requirement !

File: addons/website_crm/models/website_visitor.py [hunk 7]
```diff
@@ -7,12 +7,17 @@
...
-    lead_count = fields.Integer('# Leads', compute="_compute_lead_count")
+    lead_ids = fields.Many2many('crm.lead', string='Leads', groups="sales_team.group_sale_salesman")
+    lead_count = fields.Integer('# Leads', compute="_compute_lead_count", groups="sales_team.group_sale_salesman")
 
-    @api.depends('lead_ids.visitor_id')
+    @api.depends('lead_ids.visitor_ids')
     def _compute_lead_count(self):
-        page_data = self.env['crm.lead'].read_group([('visitor_id', 'in', self.ids)], ['visitor_id'], ['visitor_id'])
-        mapped_data = dict([(data['visitor_id'][0], data['visitor_id_count']) for data in page_data])
+        results = self.env['website.visitor'].search_read([('id', 'in', self.ids)], ['id', 'lead_ids'])
```
Comment (tde-banana-odoo) [inline addons/website_crm/models/website_visitor.py]: You can summarize this whole code with visitor.lead_count = len(visitor.lead_ids), and change trigger ^^ .

File: addons/website_crm/views/website_crm_lead_views.xml [hunk 8]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo><data>
+    <record id="website_crm.crm_lead_view_form" model="ir.ui.view">
+        <field name="name">crm.lead.form.inherit.website.crm</field>
```
Comment (tde-banana-odoo) [inline addons/website_crm/views/website_crm_lead_views.xml]: Guidelines !

ID: crm_lead_view_form
name: crm.lead.view.form.inherit.website.crm

File: addons/website_crm_livechat/__manifest__.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+{
+    'name': 'Lead Livechat Sessions',
+    'category': 'Website',
+    'sequence': 54,
```
Comment (tde-banana-odoo) [inline addons/website_crm_livechat/__manifest__.py]: Why not 57 or 78 ?

File: addons/website_crm_livechat/models/crm_lead.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+class Lead(models.Model):
+    _inherit = 'crm.lead'
+
+    visitor_sessions_count = fields.Integer('# Sessions', compute="_compute_visitor_sessions_count")
+
+    @api.depends('visitor_ids.mail_channel_ids')
+    def _compute_visitor_sessions_count(self):
+        self.flush(['visitor_ids'])
+        sql = """ SELECT l.id as lead_id, count(*) as session_count
```
Comment (tde-banana-odoo) [inline addons/website_crm_livechat/models/crm_lead.py]: Is it really costly to use the ORM ?

File: addons/website_crm_livechat/views/website_crm_lead_views.xml [hunk 11]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo><data>
+    <record id="website_crm_livechat.crm_lead_view_form" model="ir.ui.view">
+        <field name="name">crm.lead.form.inherit.website.crm.livechat</field>
```
Comment (tde-banana-odoo) [inline addons/website_crm_livechat/views/website_crm_lead_views.xml]: Naming ! (see above, or below, don't know anymore)

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-ff

**Pull request closed without merging**